### ISSUE TITLE
extension methods can break inheritance and interface implementation

### ISSUE THREAD

**Issue by MarvinHannott** [dart-lang/language#1090]
When a class extension exports a private class member, that can break extending / implementing that class:
````Dart
// A.dart
class A<T>{
  T _a;
}
extension ExtA on A<int> {
  int get a => _a;
}

// B.dart
class B<T> implements A<T> {} 

// main.dart
void main {
  final b = B<int>();
  b.a; // NoSuchMethodError
}
````
Thanks to `ExtA`, `b` also has `int get a`, but it fails with `NoSuchMethodError`, because `B` doesn't have the private member `_a`. This is very surprising and leads to accidental breaking of code. This isn't catched by Dartanalyzer. 

It gets even more confusing, because this example only crashes if `b` is a `B<int>`. Any other generic [T] would work fine. The implicit nature of extensions really works to our disadvantage here.

**Comment by leafpetersen** [comment]
I don't think this is really an issue with extensions per se, but about the problems of implementing interfaces with private members (which, to be clear, is a footgun).  The same problem arises without extensions as follows:


```dart
// A.dart
class A<T>{
  T _a;
}

class C<T> extends A<T> {
  void swap(A<T> other) {
     T x = other._a;
      other._a = _a;
      _a = x;
  }
}
int getA(A<int> a) => a._a;
```

```
// B.dart
import 'b.dart';
class B<T> implements A<T> {} 

// main.dart
void main() {
  final b = B<int>();
  getA(b); // NoSuchMethodError
  final c = C<int>();
  c.swap(b);  // NoSuchMethodError
}
```

It's true that extensions provide another way to hit this, but the underlying issue is the same.

**Comment by MarvinHannott** [comment]
You are right, in hindsight this seems obvious. Took me very much by surprise anyway. And I think your example would take anybody by surprise too. I mean, the fact this footgun even exists feels ridiculous. Why exactly was the `interface` keyword removed from Dart?

**Comment by lrhn** [comment]
An `interface` declaration wouldn't help you here either. Anything you can do with `interface`, you can also do with `abstract class`, and it won't stop someone from implementing the class anyway.

The thing you need to prevent the problem is anyone *implementing* the interface which contains the private member, and `interface` is all about being implementable.

The red flag here is that you are accessing a private member on an object that isn't guaranteed to *extend* the class declaring the private member. Normally that happens when you get such an object as a parameter, but the `this` of an *extension* is really a kind of parameter too, it's only constrained by having the right interface.

**Comment by leafpetersen** [comment]
> An `interface` declaration wouldn't help you here either. Anything you can do with `interface`, you can also do with `abstract class`, and it won't stop someone from implementing the class anyway.

I don't really think this is a fair summary.  I mean, yes, if you chose to introduce `interface` and allowed it to have private members, and still allowed classes to be implemented, then yes, abstract classes and interfaces would be equivalent.  But I think implicit in the question is: why not make different choices here?

**Comment by lrhn** [comment]
True, if introducing `interface` also meant that you can't `implement` a class, then anything which has the class' type must necessarily extend the class and have all the same private members.
So, it's not so much the introduction of `interface` as it is the prohibition on implementing classes that would come with it.

Using an `interface` declaration wouldn't help this problem in any way that an `abstract class` declaration won't also do.
